### PR TITLE
Coding-of-any-collections

### PR DESCRIPTION
**Pull request by GaroRobe**
Expands ser/de for any vector-compatible container (begin/end, resizable, random-accessable) - deque in particular.

Improved tests accordingly.

### CODE REVIEW

**Review by igor-egorov — APPROVED** [review]
lgtm

File: src/scale_encoder_stream.cpp [hunk 1]
```diff
@@ -10,144 +10,145 @@
...
-    encodeSecondCategory(value.convert_to<uint16_t>(), out);
-    return;
-  }
-
-  if (value < compact::EncodingCategoryLimits::kMinBigInteger) {
-    encodeThirdCategory(value.convert_to<uint32_t>(), out);
-    return;
+  namespace {
+    // must not use these functions outside encodeInteger
+    inline void encodeFirstCategory(uint8_t value, ScaleEncoderStream &out) {
```
[the remark this answers is not in the document]
Comment (GaroRobe) [reply src/scale_encoder_stream.cpp:15]: I don't know - it always was.

File: include/scale/scale_decoder_stream.hpp [hunk 2]
```diff
@@ -168,15 +168,17 @@ namespace scale {
...
      * @tparam T item type
-     * @param v reference to vector
+     * @param v reference to container
      * @return reference to stream
      */
-    template <class T>
-    ScaleDecoderStream &operator>>(std::vector<T> &v) {
+    template <
+        class C,
+        typename T = std::decay_t<decltype(*std::begin(std::declval<C>()))>>
```
[the remark this answers is not in the document]
Comment (GaroRobe) [reply include/scale/scale_decoder_stream.hpp]: Prettified it using trait is_iterable

File: src/scale_encoder_stream.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (GaroRobe) [reply src/scale_encoder_stream.cpp:15]: Removed

**Review by igor-egorov — APPROVED** [review]

File: test/scale_collection_test.cpp [hunk 3]
```diff
@@ -233,3 +337,22 @@ TEST(Scale, DISABLED_encodeVeryLongCollectionUint8) {
...
+  ScaleEncoderStream s;
+  ASSERT_NO_THROW((s << collection));
+  auto &&out = s.to_vector();
+
+  auto stream = ScaleDecoderStream(gsl::make_span(out));
+  std::map<uint32_t, uint32_t> decoded;
+  stream >> decoded;
+  ASSERT_TRUE(std::equal(
+      decoded.begin(), decoded.end(), collection.begin(), collection.end()));
+}
```
Comment (igor-egorov) [inline test/scale_collection_test.cpp:358]: pls add line break to the end.
IDE can be configured to do so automatically on save

**Pull request merged**